### DISCUSSION TITLE
getServerSideProps context type

### DISCUSSION THREAD

**Discussion by pitahashi** [vercel/next.js#53198] [Help] [answered]
### Summary

About `context` type passed `getServerSideProps` arguments.
I thought that should be passed `GetServerSidePropsContext` type, however it does not seemed to exist in Next.js docs.
```ts
/**
* index.d.ts
*/
export type GetServerSidePropsContext<
  Q extends ParsedUrlQuery = ParsedUrlQuery,
  D extends PreviewData = PreviewData
> = {
  req: IncomingMessage & {
    cookies: NextApiRequestCookies
  }
  res: ServerResponse
  params?: Q
  query: ParsedUrlQuery
  preview?: boolean
  previewData?: D
  resolvedUrl: string
  locale?: string
  locales?: string[]
  defaultLocale?: string
}

/**
* use case
*/
export const getServerSideProps = async (context: GetServerSidePropsContext) => {
  ...
}

```
What should we refer to information used types in that?

version
Next.js 12.3.4

**Comment by icyJoseph** [comment]
What do you mean? There's various ways to type this, one valid way is as you have it.

Another is to type the entire function: https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props

You can import `GetServerSidePropsContext` from `next`.

```ts
import type { GetServerSideProps, InferGetServerSidePropsType, GetServerSidePropsContext } from 'next';
```

So you can do:

```ts
import type { InferGetServerSidePropsType, GetServerSideProps } from 'next'
 
type Repo = {
  name: string
  stargazers_count: number
}
 
export const getServerSideProps: GetServerSideProps<{
  repo: Repo
}> = async (ctx) => {
  // do something with ctx
  const res = await fetch('https://api.github.com/repos/vercel/next.js')
  const repo = await res.json()
  return { props: { repo } }
}
```

This approach also types the return type of the function you assign to `getServerSideProps`. Which means it also handles, notFound, redirect, revalidate etc.

Or you could just do:

```ts
import type { GetServerSidePropsContext } from 'next'
 
type Repo = {
  name: string
  stargazers_count: number
}
 
export const getServerSideProps = async (ctx: GetServerSidePropsContext) => {
  // do something with ctx
  const res = await fetch('https://api.github.com/repos/vercel/next.js')
  const repo = await res.json()
  return { props: { repo } }
}
```

However in this case, you could return whatever, and TS will infer that, but it won't be checked against what next.js expects.

**Reply by pitahashi** [reply]
Thanks for comment. It means below.
1. What is types of getServerSideProps　context
2. Where should I check that type information 

Forgive me basic question.

**Reply by icyJoseph** [reply]
Do you mean this? https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props#context-parameter

You can also explore them in your text editor, but perhaps what you want is the documentation for them.

**Reply by pitahashi** [reply]
`context` 'type' means type for using on Typescript.

Generally, should I look at index.d.ts to check that? If it is possible I would like to see type docs.

**Reply by icyJoseph** [reply, marked as answer]
There's not type docs, you can just see that in your editor. They are property documented in the link I gave above though.

- https://github.com/vercel/next.js/blob/f389d815fd8730a7b952ed73ca61184c7a00f569/packages/next/types/index.d.ts#L242

You can see, it links back to the documentation page.

**Reply by pitahashi** [reply]
solved, thanks you for your answer.
